test/Test/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,525 @@
+                                   (entryCountBlock entryCount)
+
+prop_partialKeySizeBlockMakesDeserialisationFail :: [SerialisedKey]
+                                                 -> Word8
+                                                 -> NumEntries
+                                                 -> Bool
+prop_partialKeySizeBlockMakesDeserialisationFail lastKeys
+                                                 partialKeySizeByte
+                                                 entryCount
+    = isLeft $
+      fromSBS $
+      potentialSerialisedIndex
+          testedTypeAndVersionBlock
+          (lastKeysBlocks lastKeys ++ [Primitive.singleton partialKeySizeByte])
+          (entryCountBlock entryCount)
+
+prop_partialKeyBlockMakesDeserialisationFail :: [SerialisedKey]
+                                             -> Small Word16
+                                             -> Primitive.Vector Word8
+                                             -> NumEntries
+                                             -> Property
+prop_partialKeyBlockMakesDeserialisationFail lastKeys
+                                             (Small statedSize)
+                                             partialKeyBlock
+                                             entryCount
+    = fromIntegral statedSize > Primitive.length partialKeyBlock ==>
+      isLeft (fromSBS input)
+    where
+
+    statedSizeBlock :: Primitive.Vector Word8
+    statedSizeBlock = byteVectorFromPrim statedSize
+
+    input :: ShortByteString
+    input = potentialSerialisedIndex
+                testedTypeAndVersionBlock
+                (lastKeysBlocks lastKeys ++ [statedSizeBlock, partialKeyBlock])
+                (entryCountBlock entryCount)
+
+-- ** Incremental construction
+
+prop_incrementalIndexConstructionWorks :: [Append] -> Property
+prop_incrementalIndexConstructionWorks appends
+    = fst (incrementalConstruction appends) === indexFromAppends appends
+
+prop_incrementalSerialisedKeyListConstructionWorks :: [Append] -> Property
+prop_incrementalSerialisedKeyListConstructionWorks appends
+    = snd (incrementalConstruction appends) === lastKeysBlockFromAppends appends
+
+-- * Test case generation and shrinking
+
+instance Arbitrary NumEntries where
+
+    arbitrary = NumEntries <$> (getNonNegative <$> arbitrary)
+                               `suchThat`
+                               (toInteger >>> (< 2 ^ (64 :: Int)))
+
+    -- This shrinker relies on 'Int' values not becoming greater by shrinking.
+    shrink = shrinkMap (NumEntries . getNonNegative)
+                       (\ (NumEntries entryCount) -> NonNegative entryCount)
+
+instance Arbitrary SerialisedKey where
+
+    arbitrary = SerialisedKey' <$>
+                arbitrary `suchThat` (Primitive.length >>> (< 0x10000))
+
+    -- This shrinker relies on lists not becoming longer by shrinking.
+    shrink = shrinkMap SerialisedKey' (\ (SerialisedKey' bytes) -> bytes)
```

Comment:
We already have these instances in `src-extras/../Generators.hs`. The distribution for the second one might be a bit different, but I don't think it's a problem.

Reply:
Checking the corresponding code, I see that the `Arbitrary` instantiations of `NumEntries` and `SerialisedKey` reuse the instantiations of `Int` and `RawBytes`, respectively. However, what does the `Arbitrary` instantiation of `RawBytes` do? I didn’t find any documentation on this, and I must admit that the code looked complicated enough to discourage me from trying to understand it. Could you perhaps provide me with some sort of informal specification of the generator and the shrinker for `RawBytes`?

Reply:
Currently, the builds fail, because I rebased this pull request onto #406. The problem that makes them fail is that the `Arbitrary` instantiations of `NumEntries` and `SerialisedKey` from `Database.LSMTree.Extras.Generators` are now imported and clash with my own instantiations. I understand that you would like me to use the instantiations from `Database.LSMTree.Extras.Generators` anyhow, but I currently don’t want to remove my own instantiations, as there are still questions that should be answered before.

Beside the above question regarding `RawBytes`, there is also the question how one should ensure that arbitrary values of `NumPages` fit into 64 bits and arbitary serialized keys have lengths that fit into 16 bits. How do the tests for the compact index address this? Are they just based on the assumption that, with the Quick Check versions and processor architectures used, nothing will go wrong?

Reply:
> Could you perhaps provide me with some sort of informal specification of the generator and the shrinker for `RawBytes`?

At it's core it's a sequence of bytes, but the type has some extra cases that could be handled incorrectly in the code, so the tests make sure to cover them:

1. the underlying byte array being pinned or unpinned
2. only using a slice of out of a byte array

So the tests generate all these cases and also have an additional way of shrinking by just taking a smaller slice of the same byte array.

> how one should ensure that arbitrary values of NumPages fit into 64 bits and arbitary serialized keys have lengths that fit into 16 bits

Ints fit into 64 bits. For the length of serialised keys it's not a straightforward, but it depends on `QC.listof`, where the length is bounded by the size parameter (I think up to 100 by default?). So there shouldn't be issues with `SerialisedKey` being too long, but maybe one even wants to allow longer ones?

Either way, if you want to have local Arbitrary instances, it's probably a good idea to define a newtype and provide the instance for that, so you don't need orphan instances and don't have the risk of conflicting instances.

Reply:
Thanks for the explanation of the `Arbitrary RawBytes` instance. Makes sense to me. 🙂

Reply:
I think that there’s no real guarantee that `Int` values fit into 64 bits. From what I remember, the Haskell Report only specifies that `Int` covers at least the interval [−2²⁷ + 1, +2²⁷ − 1]. I can imagine that there’s no architecture currently on which GHC would actually use more than the 64-bit range for `Int`, but what about all those future 128-bit architectures? 😉

Reply:
I think I’ll use the instantiations from `Database.LSMTree.Extras.Generators` for both `NumPages` and `SerialisedKey` and add comments that specify the assumptions I’m making for the tests being correct: non-negative `Int` values being representable as 64-bit words and key lengths being representable as 16-bit words.

Reply:
I’ve made the corresponding changes in f9fb2816cc8b88aa4dca701885857fed58766b56.